design-proposals/shadow-node.md

```diff
@@ -0,0 +1,86 @@
+# Overview
+
+A reported [Advisory](https://github.com/kubevirt/kubevirt/security/advisories/GHSA-cp96-jpmq-xrr2) is outlining how of our virt-handler component can be abused to escalate local privileges when a node, running virt-handler, is compromised. A flow and more details can be found on reported [issue](https://github.com/kubevirt/kubevirt/issues/9109).
+
+This proposal is outlining mitigation in a form of internal change.
+
+## Motivation
+
+Kubevirt should be secure by default. While mitigation is available, it is not part of Kubevirt.
+
+## Goals
+
+Mitigate the advisory by default.
+
+## Non Goals
+
+## Definition of Users
+
+Not user-facing change
+
+## User Stories
+
+Not user-facing change
+
+## Repos
+
+Kubevirt/Kubevirt
+
+# Design
```

Comment:
One alternative design that was proposed the sig-scale call, that may solve the same problem is as follows:

1. Introduce a new controller, lets call it token-distributor-controller.
2. The token-distributor-controller has a RBAC Template CR that creates a unique ClusterRole, ClusterRolebinding and ServiceAccount for each virt-handler pod. Example of the template from kubevirt-handler clusterrole

```
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - list
  - watch
  - get
- apiGroups:
  - ""
  resources:
  - nodes
  resourceNames:
  - <node-name>
  verbs:
  - patch
```
3. The token-distributor-controller will then create a token for the pod via `TokenRequest` API [1][2] pod on the node and deliver the token on the node. 
4. The virt-handler pod will use a local volume mount to receive the token. An init container could be used to make sure the file is delivered by token-distributor-controller
5. Virt-handler will start with restricted permissions that could be carefully modified so that it will work on all the objects that are related to the node

This could be a generic solution that could help all daemonsets that are deployed with privileged RBAC and could be a separate project all together.  

Href:
1. https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount 
2. https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/

Reply:
I don't think that this is addressing the probelm fully. There are two thing which need to be fixed:

 1. being able to modify other nodes (that can be addressed like above)
 2. Not being able to modify non-kubevirt-owned fields on the node (this includes not being able to change the spec or any label/annotation which is not stricty kubevirt owned)
 
 
 We could definitely create unique roles for each node, and in addition deploy a node webhook which would ensure that a service account only, that is also what the apiserver does internally for each kubelet, but that has the severe disadvantage that we would now need a **required** webhook on a core k8s resource. If our node webhook is ever unavailable, no kubelet, or anythig else can actually continue to modify the node. That's a no-go.

THere are other positive side-effects if we fix it with the "shadow nodes":

 * heartbeat can be moved to the shadow node, meaning we create much less load on all the controllers watching nodes.
  * we finally have a clear api to express our node-needs and don't need to press our API in annotations and labels on nodes.
 
 
 One may now think that now virt-controller still can modify all nodes, and that is true. However now we only have one component which can do that instead of n, further a `token-distributor-controller` controller which would create those rbac roles has to have the exact same permissions as well, otherwise it would not be able to create the roles in the first place ...

Reply:
@rmohr thanks for summarizing the reasons. 

@xpivarc if it is helpful, please add this to alternative designs considered for future us to refer back at.

Reply:
Can we includes this in the gola section?

@RamLavi @xpivarc ?

Reply:
[alaypatel07](https://github.com/alaypatel07) thanks for bringing this up. 
Added the alternative to the proposal body.
Added to the goals.

Reply:
@rmohr we have to address a third element

3. Prevent that the daemonset can modify other KubeVirt owned CRs related to other nodes

IOW 3. Restrict the damemonset to only modify KubeVirt owned CRs related to it's own node

Reply:
Yes that third element is true, but I would qualify it as a follow-up step, not needed to address the CVE. Now virt-handler cannot blindly modify the node topology anymore. It can only modify hardware capability availability. It can not lure in privileged or sensitive workloads anymore by just applying labels which they are waiting for, or by making a node a control-plane node.

The follow up would be either cert based auth for virt-handler, or maybe we can still use tokens and get the IP in the webhooks for verification. Something like this ...

Let's add it to the doc as follow-up.

Reply:
@fabiand AFAIK the only kubevirt CR virt-handler can update is `virtualmachineinstances`. Is this what you are referring to?

@rmohr I will add a `Follow-up Steps` section in the end of the doc and include:
- (the one just suggested) Restrict the virt-handler daemonset to only modify KubeVirt owned CRs related to it's own node
- Restrict the virt-handler daemonse from being able to update/patch other nodes

Is that alright with you?

Reply:
@RamLavi yes - if this is the only one, then yes. And: generally speaking: handler should just be able to touch anything which a) it should be permitted to touch and b) only if it's own it's "own node"

Reply:
ACK. DONE.